---
layout: post
title: "SNP vs SNV"
description: ""
category: 
tags: []
---
{% include JB/setup %}

Ian Holmes has a [twitter poll][poll] right now on the use of "SNP"
(single-nucleotide polymorphism) versus "SNV" (single-nucleotide variant). I
have been bugged by the two terminologies for years, so I decided to write a
blog post on it. Personally, **I use "SNP" for germline events and "SNV" for
somatic events**, but I understand others think differently. Here are my
thoughts.

The [wiki page for SNP][snp-wiki] defines a SNP as a nucleotide change "that is
present in a sufficiently large fraction of the population (e.g. 1% or more)".
However, **such a frequency-based definition is not actionable in practice**.
Allele frequency varies a lot across populations. Due to genetic drift and
selection, an allele at 5% frequency in African may be absent from the rest of
the world. Is this a SNP or not? Furthermore, the observed allele frequency 
fluctuates with sampling and the sample size. An allele at 2% frequency in the
1000 Genomes Project (1KG) may become 0.5% in gnomAD. Is this a SNP or not? If
it is impractical to set a frequency threshold, the definition of SNP shouldn't
require a frequency threshold.

**Historically, we have been using "SNP" without a frequency threshold for
decades**. If you search word "SNP" in the [landmark paper on the Human Genome
Project][hgp-paper] in 2001, you can find 45 instances. With data produced at
that time, we had little information on frequency but we called observed
substituions as SNPs anyway. Similarly, there are 28 instances of "SNP" in the
[final 1KG paper][1kg-paper], including one in the abstract. In 1KG, we have
observed many substitutions at &lt;1% but we still called them as SNPs.
In these papers, a SNP simply refers to a germline substitution.

**"SNV" is a much more recent terminology**. I first saw "SNV" in the [SNVmix
paper][snvmix-paper] in the context of tumor mutation calling (I reviewed it).
That was 2010.  According to [a PubMed search][pubmed-search], few papers were
published with "SNV" in the abstract before that and early uses of "SNV" mostly
focused on tumor data as well. This includes the popular [VarScan2
paper][vs2-paper]. People coined up "SNV" for somatic mutations because SNP has
been reserved for germline events. "SNV" may sound more general than "SNP", but
concepts in genetics should not be taken literally. What matters more is the
historical uses. There are simiarly confusing terminologies like VNTR and CNV
vs CNA, which I will not explain in detail here.

It is already too late to regulate the use of SNP and SNV. In practice, just
beware that the definition of SNP and SNV may vary between researchers. When
in a conversation you are not sure what SNP/SNV refers to, ask for a
clarification.

**Postscript:** I personally avoid "SNV" in my work due to its inconsistent
uses in the past. When I want to describe a somatic event, I use "somatic SNV"
or "sSNV" in brief.

[poll]: https://twitter.com/ianholmes/status/1371523573861339141
[snp-wiki]: https://en.wikipedia.org/wiki/Single-nucleotide_polymorphism
[hgp-paper]: https://www.nature.com/articles/35057062
[1kg-paper]: https://www.nature.com/articles/nature15393
[snvmix-paper]: https://pubmed.ncbi.nlm.nih.gov/20130035/
[pubmed-search]: https://pubmed.ncbi.nlm.nih.gov/?term=%28"single+nucleotide+variants"%5Btiab%5D+OR+"single+nucleotide+variant"%5Btiab%5D%29+SNV%5Btiab%5D&filter=dates.1000%2F1%2F1-2011&filter=dates.1000%2F1%2F1-1980
[vs2-paper]: https://pubmed.ncbi.nlm.nih.gov/22300766/
